Adds popscope migration guide (#9872)
flutter/website: src/release/breaking-changes/popscope-with-result.md

---
title: Generic types in PopScope
description: >
  Added a generic type in PopScope class and updated the onPopInvoked
  function signature.
---

## Summary

> Added a generic type in [`PopScope`][] class and updated the [`onPopInvoked`][]
> function signature to take in an additional result as an position parameter.

Reviewer: ```suggestion
Added a `Route`'s generic type to the [`PopScope`][] class
and updated the [`onPopInvoked`][] function signature
to take this type as a positional parameter.
```

## Context

> Previously, PopScope hadn't haven a way to access the pop result when `onPopInvoked`
> is called. To add support for accessing pop result, the generic type is added to
> PopScope class so that the `onPopInvoked` access the type-safe result.

Reviewer: ```suggestion
Previously, the `onPopInvoked` method in `PopScope`
didn't have a type-safe way to access the
`Route`'s pop result. This change addresses that issue.
```

## Description of change

> Added a generic type in `PopScope` class and updated the `onPopInvoked`
> function signature to take in an additional result as an position parameter.

Reviewer: ```suggestion
Added a generic type in the `PopScope` class and updated
the `onPopInvoked` method to take the type as a positional parameter.
```

## Migration guide

Code before migration

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      navigatorKey: nav,
      home: PopScope(
        canPop: false,
        onPopInvoked: (bool didPop) {
          if (didPop) {
            return;
          }
          launchConfirmationDialog();
        },
        child: MyWidget(),
      ),
    ),
  );
}
```

Code after migration
```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      navigatorKey: nav,
      home: PopScope<Object?>(
        canPop: false,
        onPopInvoked: (bool didPop, Object? result) {
          if (didPop) {
            return;
          }
          launchConfirmationDialog();
        },
        child: MyWidget(),
      ),
    ),
  );
}
```

The generic type should match the generic type of the [`Route`] the `PopScope` is in.

> If the route uses `int` as its generic type, consider use `PopScope<int>`.

Reviewer: ```suggestion
If the route uses `int` as its generic type, consider using `PopScope<int>`.
```

> If the widget are shared across multiple routes with different types, one can

Reviewer: ```suggestion
If widgets are shared across multiple routes with different types,
```

use `PopScope<Object?>` to catch all possible types.

## Timeline

> Landed in version: TBD

Reviewer: Please fill this out.

Author: will update after the change is merged
